Report reasons for not caching

wolfs opened the pull request.

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/TaskOutputsInternal.java:
```diff
@@ -47,9 +49,8 @@
+     * @param reasons Reasons for disabling the cache are added to this list.
      */
-    boolean isCacheAllowed();
+    boolean isCacheEnabled(List<TaskCachingDisabledReason> reasons);
```
This looks a little funky.

Maybe something like...

```
TaskOutputCaching getTaskOutputCaching()

interface TaskOutputCaching {
   boolean isCachedAllowed()
   boolean isCachedEnabled()
   List<TaskCachingDisabledReason> getReasons()
}
```

big-guy replied:
This looks a little funky.

Maybe something like...

```
TaskOutputCaching getTaskOutputCaching()

interface TaskOutputCaching {
   boolean isCachedAllowed()
   boolean isCachedEnabled()
   List<TaskCachingDisabledReason> getReasons()
}
```

big-guy replied:
Looking at how this is used... do we have any strong reason to make the reasons something other than a String?

wolfs replied:
At some point it felt like I wanted to distinguish between the task not being annotated at all and some other reasons like using `doNotCacheIf`. Looking at this closer I think we should be fine with having just a list of Strings. It feels like a good idea to expose the interface `TaskOutputCaching` to the listener? The other option would be to not use a listener at all but to have the client call `task.getOutputs().getTaskOutputCaching()`. WDYT?

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/AbstractTaskOutputsDeprecatingTaskPropertyBuilder.java:
```diff
@@ -52,11 +52,21 @@ public void cacheIf(Spec<? super Task> spec) {
+
+    @Override
     public void doNotCacheIf(Spec<? super Task> spec) {
         getTaskOutputs("doNotCacheIf(Spec").doNotCacheIf(spec);
```
This was pre-existing.  Missing closing )

wolfs replied:
Fixed via https://github.com/gradle/gradle/commit/3ad75001b0e6bd2b69b40128bf994c7ec9384fb1

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -43,8 +47,8 @@
-    private AndSpec<TaskInternal> cacheIfSpec = AndSpec.empty();
-    private OrSpec<TaskInternal> doNotCacheIfSpec = OrSpec.empty();
+    private List<SpecWithDescription<TaskInternal>> cacheIfSpecs = new ArrayList<SpecWithDescription<TaskInternal>>();
+    private List<SpecWithDescription<TaskInternal>> doNotCacheIfSpecs = new ArrayList<SpecWithDescription<TaskInternal>>();
```
Maybe we can use this interface that Cedric's adding: https://github.com/gradle/gradle/pull/1178/files#diff-781b81824187b0f1558206a030f0c4dc ?

wolfs replied:
I am now using `Describable`.

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -87,14 +91,37 @@ public void run() {
+            return false;
+        }
+        if (!hasDeclaredOutputs()) {
+            reasons.add(new DefaultTaskCachingDisabledReason("No outputs declared"));
```
should this be returning false?

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -87,14 +91,37 @@ public void run() {
 
     @Override
-    public boolean isCacheAllowed() {
+    public boolean isCacheEnabled(List<TaskCachingDisabledReason> reasons) {
```
Should we do something to cache the result of this? i.e., so we don't re-evaluate the Specs multiple times?

wolfs replied:
Currently, we are not re-evaluating the spec multiple times - we are only calling it from the `SkipCachedTaskExecuter`. The result of evaluating the spec could change over time and anyone can call this method via `task.outputs.taskCaching`.
Should we add some validation and only execute the method once or leave it as is?

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -103,18 +130,28 @@ public boolean isCacheAllowed() {
         taskMutator.mutate("TaskOutputs.cacheIf(Spec)", new Runnable() {
             public void run() {
-                cacheIfSpec = cacheIfSpec.and(spec);
+                cacheIfSpecs.add(new SpecWithDescription<TaskInternal>(spec, message));
```
Maybe add something...

`Spec Specs.describe(String message, Spec spec)`

So we could do something like...

`cacheIfSpec = cacheIfSpec.and(Specs.describe(message, spec))`

And we could make the `CompositeSpec` build a description for the whole chain?

wolfs replied:
Do you think adding this API to Specs is worth it? We currently only record reasons for cacheIf Specs which yield false and doNotCacheIf Specs which yield true. That would mean that either `AndSpec` and `OrSpec` would have to gather the descriptions for specs which yield true and which yield false. I would not add that complexity now but revisit it if we also want to report reasons why up to date checks for a task are disabled.
WDYT?

big-guy commented on subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java:
```diff
@@ -139,9 +147,18 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
+    private static TaskExecuter createSkipCachedExecuterIfNecessary(StartParameter startParameter, BuildCacheConfigurationInternal buildCacheConfiguration, TaskOutputPacker packer, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskOutputOriginFactory taskOutputOriginFactory, ListenerManager listenerManager, TaskExecuter delegate) {
         if (startParameter.isTaskOutputCacheEnabled()) {
-            return new SkipCachedTaskExecuter(taskOutputOriginFactory, buildCacheConfiguration, packer, taskOutputsGenerationListener, delegate);
+            TaskCachingReasonsListener cachingReasonsListener = listenerManager.getBroadcaster(TaskCachingReasonsListener.class);
```
I dislike having logic in the services classes...

Maybe we can move this to another TaskExecuter that handles the listener?

wolfs replied:
Where are we currently adding all the listeners for logging? Does this happen in `DefaultGradleLauncher`? We could add the listener there and only add the executer here.

lptr replied:
I agree the logic would be better if it wasn't in the service registry. How about moving the listener to its own service instead, and inject it as such? Adding another executer just for this seems to be the wrong abstraction level to handle this.

big-guy replied:
We still have to have something that glues together the listener instance with the listener manager.

Maybe if we just move it up into the `createTaskExecuter` above and unconditionally create the listener, that would be slightly better.

wolfs commented:
@big-guy I introduced `TaskCaching` on `TaskOutputsInternal`. It feels to me that `getTaskCaching` actually wants to live on the task interface itself. What do you think? Do you think the listener interface is good now or should we add something to `TaskStateInternal` or even `Task` to obtain task caching and ditch the listener altogether?

wolfs commented:
@big-guy @lptr I think that PR is nearly ready to be merged. Do you have some other comments?

lptr commented on subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedTaskExecutionIntegrationTest.groovy:
```diff
@@ -305,6 +305,22 @@ class CachedTaskExecutionIntegrationTest extends AbstractIntegrationSpec impleme
+        then:
+        output.contains "The doNotCacheIf(Spec) method has been deprecated and is scheduled to be removed in Gradle 4.0. Please use the doNotCacheIf(String, Spec) method instead."
+    }
+
```
Should we do the same with `cacheIf(Spec)`?

wolfs replied:
I was thinking not to deprecate `cacheIf` without a reason since it can be used to turn caching on:
```groovy
cacheIf {
   true
}
```
If we required a reason in that case (could be as easy as "Turn caching on"), then the reason would currently not be printed - so I was thinking why should we force that on the user. What is your take on that?

big-guy replied:
I think it's fine to leave this non-deprecated.

lptr replied:
Makes sense, thanks.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/TaskOutputsInternal.java:
```diff
@@ -37,19 +37,12 @@
+     * Yields information about the cacheability of the outputs.
      */
-    boolean isCacheEnabled();
+    TaskCaching getTaskCaching();
```
I would probably call the type `TaskOutputCaching` and the method `getCaching()` as `Task` and `Output` is already in the name of the type the method is on.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/TaskOutputsInternal.java:
```diff
@@ -37,19 +37,12 @@
-     * {@literal @}{@link org.gradle.api.tasks.OutputFiles} or
-     * {@literal @}{@link org.gradle.api.tasks.OutputDirectories}; or {@code true} otherwise.
-     */
-    boolean isCacheAllowed();
```
\o/

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/DescribableSpec.java:
```diff
@@ -0,0 +1,40 @@
+import org.gradle.api.Describable;
+import org.gradle.api.specs.Spec;
+
+public class DescribableSpec<T> implements Describable, Spec<T> {
```
WDYT about `SelfDescribingSpec` as a name? Easier on the tongue!

wolfs replied:
I changed it to `SelfDescribingSpec`.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskCachingReasonsListener.java:
```diff
@@ -0,0 +1,24 @@
+import org.gradle.api.internal.TaskCaching;
+
+public interface TaskCachingReasonsListener {
+    void taskCacheable(Task task, TaskCaching cacheable);
```
This method name doesn't help me understand when this method should be called, or what happens when it gets called. Some Javadoc would help, but the best would be to find a name that describes the role of the method better. Something with a verb would definitely improve matters.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java:
```diff
@@ -121,6 +127,7 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
                                             packer,
                                             taskOutputsGenerationListener,
                                             taskOutputOriginFactory,
+                                            listenerManager,
```
Can we handle the new listener the same way we handle `taskOutputsGenerationListener`? Passing the whole manager seems unnecessary here.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuter.java:
```diff
@@ -47,81 +48,69 @@
+            TaskCaching taskCaching = taskOutputs.getTaskCaching();
+            cacheable = taskCaching.isCacheable();
+            state.setCacheable(cacheable);
+            taskCachingReasonsListener.taskCacheable(task, taskCaching);
```
Since we are passing the whole `task` object anyway, adding the separate `TaskCaching` type seems a bit redundant.

wolfs replied:
That would mean that we could drop the listener altogether and have a client use a `TaskExecutionListener`. By listening to `beforeExecute(Task)` the client would be able to obtain `task.outputs.caching`. I feels to me that we probably want to do that given that I cannot come up with a good name for the method of the listener. Only something like `taskCachingEvaluated` would come to mind. We also could add the whole immutable `TaskOutputCaching` object to `TaskStateInternal` removing `TaskStateInternal.isCacheable()` in the process.
What is your take on this?

lptr commented on subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskOutputsTest.groovy:
```diff
@@ -241,53 +241,96 @@ class DefaultTaskOutputsTest extends Specification {
+        then:
+        !outputs.taskCaching.cacheable
+        outputs.taskCaching.disabledReason == "Declares multiple output files for a single output property via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`"
+
```
Should we have a test for the value of `disabledReason` when the task _is_ cacheable?

wolfs replied:
I added a test.

lptr commented on subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/execution/SkipCachedTaskExecuterTest.groovy:
```diff
@@ -244,18 +212,14 @@ class SkipCachedTaskExecuterTest extends Specification {
         ex.cause instanceof RuntimeException
         ex.cause.message == "Bad cacheIf() clause"
```
We could also report the description of the `cacheIf()` clause here, if there's any. That would help debugging a lot. Not sure how frequently this would be used, though.

wolfs replied:
The description of the `cacheIf` and `doNotCacheIf` specs are now reported.

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputCaching.java:
```diff
@@ -0,0 +1,44 @@
+    private final String disabledReason;
+
+    static final DefaultTaskOutputCaching CACHEABLE = new DefaultTaskOutputCaching(true, null);
+    static DefaultTaskOutputCaching notCacheable(String disabledReason) {
```
Nit, return `TaskOutputCaching`?

wolfs replied:
fixed

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputCaching.java:
```diff
@@ -0,0 +1,44 @@
+
+    static final DefaultTaskOutputCaching CACHEABLE = new DefaultTaskOutputCaching(true, null);
+    static DefaultTaskOutputCaching notCacheable(String disabledReason) {
+        return new DefaultTaskOutputCaching(false, disabledReason);
```
maybe we should enforce that `disabledReason` cannot be null here?

big-guy replied:
Or enforce that `disabledReason` cannot be null if `cacheable` is false in the constructor?

wolfs replied:
fixed.

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -86,35 +89,57 @@ public void run() {
         for (TaskPropertySpec spec : getFileProperties()) {
             if (spec instanceof NonCacheableTaskOutputPropertySpec) {
-                return false;
+                return DefaultTaskOutputCaching.notCacheable("Declares multiple output files for a single output property via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`");
```
could this include the property name?

wolfs replied:
fixed.

big-guy commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java:
```diff
@@ -86,35 +89,57 @@ public void run() {
+        }
+        for (SelfDescribingSpec<TaskInternal> selfDescribingSpec : cacheIfSpecs) {
+            if (!selfDescribingSpec.isSatisfiedBy(task)) {
+                return DefaultTaskOutputCaching.notCacheable(selfDescribingSpec.getDisplayName());
```
One weird thing about this.  The descriptions for `doNotCacheIf` specs are the reason for why the task isn't cacheable.  It would seem that the descriptions for `cacheIf` specs should be the reason for why the task *is* cacheable, but we're using it for the opposite here.

Maybe with something like:

```
myTask {
   cacheIf("on CI") {
       System.getenv("CI") != null
   }
}
```

We could say the reason is "Not caching task ':myTask': cacheIf not satisfied 'on CI'".

I'm not sure this is any better than what you already have.  I think it's clear enough that the reason is the reason for disabling caching.

wolfs replied:
`cacheIf` now takes an enabled reason while `doNotCacheIf` takes a disabled reason.

breskeby commented:
From BuildScan perspective it's great to have the not cacheable reason in TaskStateInternal. That makes it easy to consume from the BuildScan plugin. But strictly speaking I'm not sure this is the right place for this info Is the reason part of the state? Feels more like a meta info about the state. just my 2 cents.

etiennestuder commented:
> (I don't remember if we thought we should do this now but) it's inconsistent that we only set the cacheable status for a task when it's not up-to-date.

It would be great if you can always set it because we are going to allow filtering by that state. And it should also be correct - not just for when the task is not up-to-date.

wolfs commented:
I removed the listener and added `TaskOutputCaching` to `TaskStateInternal`. I agree that it is not part of the state but a description why the task had the given outcome. I.e. the task was executed since it is not cacheable. One could think that we want to add the reasons why a task is not up to date at some point, too. The would have something like "the task was not up to date since `upToDateWhen` returned false for reason X". It does not feel right to have this in `TaskStateInternal` - but `TaskStateInternal` would probably be `TaskOutcome` at some point. And then having reasons why the `TaskOutcome` was `EXECUTED` and not `UP-TO-DATE` would make sense.
In the current state of the PR the information if the task is cacheable or not is always added to `TaskStateInternal` no matter what the outcome of the task is. This means we can evaluate this information even for `UP-TO-DATE` tasks.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/service/scopes/TaskExecutionServices.java:
```diff
@@ -107,25 +111,27 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository, TaskOutp
-                                                new ExecuteActionsTaskExecuter(
-                                                    taskOutputsGenerationListener,
-                                                    listenerManager.getBroadcaster(TaskActionListener.class)
+                new TaskOutputCachingEvaluationExecuter(
```
This shouldn't be added if there is no caching enabled.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskOutputCachingEvaluationExecuter.java:
```diff
@@ -0,0 +1,49 @@
+            state.setTaskOutputCaching(taskOutputCaching);
+            if (!taskOutputCaching.isEnabled()) {
+                LOGGER.info("Not caching {}: {}", task, taskOutputCaching.getDisabledReason());
+            }
```
When task output caching is off, we should disable provide `DISABLED` as the cacheability here.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/TaskOutputCaching.java:
```diff
@@ -0,0 +1,32 @@
+
+import org.gradle.api.Nullable;
+
+public interface TaskOutputCaching {
```
I still have some "off" feeling about the name of this type. This is essentially the `TaskOutputCachingState`, no? Or maybe the `TaskOutputCachingResult`. But I think I prefer state.

wolfs replied:
I had `TaskCacheability` at some point. I describes if the outputs of the task are cacheable. In other words if caching the outputs of a task is enabled. How about `CachingTaskOutputs`?
The field describes characteristics of the task outputs with respect to caching. Do you think `State` is the right noun to describe characteristics? I feel that `State` is something that easily changes.

wolfs commented:
@lptr I addressed your comments. Could you have another look?
